src/ucp/tag/tag_send.c

```diff
@@ -291,36 +341,34 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_tag_send_sync_nb,
         goto out;
     }
 
-    req = ucp_request_get(ep->worker);
-    if (req == NULL) {
-        ret = UCS_STATUS_PTR(UCS_ERR_NO_MEMORY);
-        goto out;
+    if (param->op_attr_mask & UCP_OP_ATTR_FIELD_REQUEST) {
+        req = (ucp_request_t *)param->request - 1;
+    } else {
+        req = ucp_request_get(ep->worker);
+        if (req == NULL) {
+            ret = UCS_STATUS_PTR(UCS_ERR_NO_MEMORY);
+            goto out;
+        }
     }
 
     ucp_tag_send_req_init(req, ep, buffer, datatype, count, tag,
                           UCP_REQUEST_FLAG_SYNC);
+    if (param->op_attr_mask & UCP_OP_ATTR_FLAG_NO_ZCOPY) {
+        rma_thresh   = ucp_ep_config(ep)->tag.rndv_send_nbr.rma_thresh;
+        am_thresh    = ucp_ep_config(ep)->tag.rndv_send_nbr.am_thresh;
+    } else {
+        rma_thresh   = ucp_ep_config(ep)->tag.rndv.rma_thresh;
+        am_thresh    = ucp_ep_config(ep)->tag.rndv.am_thresh;
+    }
```

Comment:
also seems common code
maybe move it inside ucp_tag_send_req? so ucp_tag_send_req() will accept "param" as argument